Coroutines/Coroutines.md

```diff
@@ -1 +1,131 @@
+
+In the snippet below, an existing asynchronous API service that uses callbacks is wrapped into a suspending function, and it propagates the result or error using a Continuation. It’s just an example function, but the idea is there.
+```kotlin
+suspend fun <Data, Result> suspendAsyncApi(data: Data): Result =
+  suspendCancellableCoroutine { continuation ->
+    apiService.doAsyncStuff<Data, Result>(data,
+        { result -> continuation.resume(result) }, // resume with a result
+        { error -> continuation.resumeWithException(error) } // resume with an error
+    )
+  }
+```
+## CoroutineContext
+
+Coroutines always execute in some context represented by a value of the [CoroutineContext](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.coroutines/-coroutine-context/) type, defined in the Kotlin standard library.
+
+- Is a set of elements (threading policy, logging, security and transaction aspects of the coroutine execution, coroutine identity and name, etc)
+- Immutable
+- Can be composed  (see [Coroutine Context](#coroutine-context)  chapter)
+
+Conceptually, coroutine context is an indexed set of elements, where each element has a unique key. It is a mix between a set and a map. Its elements have keys like in a map, but its keys are directly associated with elements, more like in a set. 
+
+## Coroutine Scope
+
+`CoroutineScope`s confine new coroutines by providing a lifecycle-bound component that binds to a coroutine. Every coroutine builder is an extension function defined on the `CoroutineScope` type. `launch()` is an example of a coroutine builder.
+
+Example - `GlobalScope`(Avoid using it). It’s useful for top-level coroutines that operate on the entire app lifetime and aren’t bound to any lifecycle. Typically, you’d use `CoroutineScope` in your `Activity`, `Fragment` or `ViewModel` over `GlobalScope` in an Android app to control when lifecycle events occur.
+```kotlin
+class Activity {
+    private val mainScope = MainScope()
+
+    fun destroy() {
+        mainScope.cancel()
+    }
+}
+```
+Or use delegation with default factory functions:
+```kotlin
+class Activity : CoroutineScope by CoroutineScope(Dispatchers.Default) {}
+```
+## Coroutine Dispatcher
+
+Dispatchers determine what thread or thread pool the coroutine uses for execution. The dispatcher can confine a coroutine to a specific thread. It can also dispatch it to a thread pool. Less commonly, it can allow a coroutine to run unconfined, without a specific threading rule, which can be unpredictable (see [Dispatchers](#dispatchers) chapter).
+
+Common Dispatchers:
+
+- [Dispatchers.Main](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-main.html): This dispatcher confines coroutines to the main thread for UI-driven programs, like Swing, JavaFX, or Android apps. **It’s important to note that this dispatcher doesn’t work without adding an environment-specific Main dispatcher dependency in Gradle or Maven**.
```

Comment:
Let's add here an example of providing Dispatcher?
[implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.3'](https://github.com/Kotlin/kotlinx.coroutines)